### ISSUE TITLE
concurrency issue: modify component between prepare and def resolved

### ISSUE THREAD

**Issue by ged-odoo** [odoo/owl#421]

**Comment by alexkuhn** [comment]
### JS
```js
const { Component, Store } = owl;
const { useDispatch, useStore } = owl.hooks;

//---------------------------------------------------------
// Store
//---------------------------------------------------------

const actions = {
    async loadThread({ state }, threadId) {
        const thread = state.threads[threadId]
        thread.state = 'loading';
        await Promise.resolve().then(() => {}).then(() => {});
        thread.state = 'loaded';
    },
    openChatWindow({ state }, { threadId }) {
        state.chatWindows.push(threadId);
    },
};

const state = {
    chatWindows: [],
    threads: {
        1: {
            id: 1,
            isFolded: false,
            state: 'non-loaded',
        },
    },
};

const store = new Store({ actions, state });

//---------------------------------------------------------
// Message List
//---------------------------------------------------------

class MessageList extends Component {}

//---------------------------------------------------------
// Thread
//---------------------------------------------------------

class Thread extends Component {
    dispatch = useDispatch();
    thread = useStore((state, props) => state.threads[props.threadId]);

    mounted() {
        if (this.thread.state == 'non-loaded') {
           this.dispatch('loadThread', this.thread.id);
        }
    }
}

Thread.components = { MessageList };

//---------------------------------------------------------
// Chat Window
//---------------------------------------------------------

class ChatWindow extends Component {
    thread = useStore((state, props) => state.threads[props.threadId]);

    isFolded() {
        return this.thread.isFolded;
    }
}

ChatWindow.components = { Thread };

//---------------------------------------------------------
// Chat Window Manager
//---------------------------------------------------------

class ChatWindowManager extends Component {
    dispatch = useDispatch();
    chatWindows = useStore(state => state.chatWindows);
    
    _onOpenChatWindow() {
        this.dispatch('openChatWindow', { threadId: 1 });
    };
}

ChatWindowManager.components = { ChatWindow };

//---------------------------------------------------------
// App
//---------------------------------------------------------

class App extends Component {}
App.components = { ChatWindowManager };

const qweb = new owl.QWeb(TEMPLATES);
const app = new App({ qweb, store });
app.mount(document.body);
```

### XML
```xml
<templates>
  <div t-name="MessageList" class="message_list"/>
  
  <div t-name="Thread" class="thread">
    <span>Thread <t t-esc="thread.id"/> (<t t-esc="thread.state"/>)</span>
    <div t-if="thread.state !== 'loaded'">
      loading...
    </div>
    <t t-else="">
      <MessageList threadId="thread.id"/>
    </t>
  </div>
  
  <div t-name="ChatWindow" class="chat_window">
    <span>Chat Window (<t t-if="isFolded">folded</t><t t-else="">unfolded</t>)</span>
    <Thread threadId="props.threadId"/>
  </div>
  
  <div t-name="ChatWindowManager" class="chat_window_manager">
    <button t-on-click="_onOpenChatWindow">open chat window</button>
    <t t-foreach="chatWindows" t-as="threadId">
      <ChatWindow threadId="threadId" t-key="threadId"/>
    </t>
  </div>

  <div t-name="App" class="app">
    <ChatWindowManager/>
  </div>
</templates>
```

Click on "open chat window"
=> `Cannot read property 'sel' of null`

**Comment by ged-odoo** [comment]
excellent!

**Comment by aab-odoo** [comment]
@alexkuhn Does the issue still occur with the last version of owl?

**Comment by alexkuhn** [comment]
Still happens with current version, currently working on reproducible example on playground

**Comment by alexkuhn** [comment]
The test that crashed and initiated this issue is solved by #448. However, we have made more tests while keeping v0.23 (before the concurrent pipeline changes), and some of them now crash in 1.0.0-alpha2 due to concurrency issues again. #448 does not fix them.

After investigating the problem and thinking about it as a whole, **it cannot be fully solved by "cancelling" outdated renderings like this** if we keep asynchronous rendering and observable mutations.

The cause of the problem is having both _concurrency_ and _observable mutations_ at the same time. We definitely want to keep concurrency, so **the only way to solve the issue is to not mutate observable**. That way, some code during rendering may use outdated data and never crash.

To do so, we could change observable so that there is a "pseudo-mutating" part (`proxy`) and an "actual, raw and immutable" part (`ref`). The `proxy` prevents using spread operators "a la react", while `ref` is build automatically from "mutating" `proxy`.

With above solution, in addition to solving all concurrent rendering issues, this would make the behaviour of `useStore()` more consistent (basically an equality checker on received `ref` data), and this would enable some important features like observable primitive types and smarter memoization of store getters.

Don't get me wrong, I think this is a good idea to cancel useless rendering. However, this should be limited to performance reasons, not hiding a design flaw.

**Comment by ged-odoo** [comment]
Dear @alexkuhn 

I am not sure if you are trolling, but in doubt, let me answer this.

First, please do not add comments on closed issues like this (unless the specific issue is still not resolved).  This makes it hard to track, and will probably mean that we cannot ensure we properly handled everything.  Just open new issues if you have a bug or a discussion.  I will need to manually ping @aab-odoo to make sure he is aware of this.

Second, if you have other issues, it would be nice to describe them more precisely. Your comment does not help us solving them at all.

Third, your suggestion is interesting (note that it's actually another issue: #294 ), but I am sad to have to tell you that you are **wrong, on "multiple levels", for many "reasons"**.


### 1. _concurrency_ and _observable mutations_ are orthogonal properties.

One does not imply (nor require) the other. Or, to say it differently, the rendering pipeline  is not related at all to observable mutations. It does not care.
  
We needed to rewrite it to solve properly the way we handle asynchronous rendering. Not because  we hide a design flaw. Because asynchronous rendering issues are not limited to what you seem to think, but to incredibly many other situations where no observable  mutations are involved.

Canceling useless renderings is NOT for performance reason. This avoid many  corrupted situations.

Please look at all the concurrent renderings scenarios in `component.test.ts`. You will see that few (or none) of them are caused by observable mutations issues.  And all of them are about coordinating rendering across time, and across components.

### 2. Doing what you suggest will not solve all concurrent rendering issues, at all. Not even close.

See above. It looks like you only think of a specific kind of issue: shared state from some component to one of this children.

In the real world, there are a lot of other cases. Some of them can be handled by Owl (still working on this). Some others are impossible to solve at a framework level: if you use a global variable that is changing, and you have a lot of very asynchronous components, that's on you.

### 3. Your comment on `useStore` does not make much sense

How does making a strict equality check on a ref object is more consistent than checking the equality of two rev numbers? It has the exact same semantics. You gave an example that would actually not gain much from your suggestion.

### 4. [opinion] observable primitive types are not important

I assume that you are talking about the fact that the result of `useStore` should be an object or an array. I do not understand how this limitation is important to you.  This limitation does not reduce the expressivity of the system.  Also, note that Vue 3 will have the exact same limitation.

-----

Let us talk about what you are suggesting: some kind of mix between mutations/changing in place. Well, first, this is probably exactly what #294 is about. So, yes, this is clearly nice if we can make it work. And you are welcome to work on this. I am not sure how hard/easy it would be, but I agree.

Also, doing it would not change much about the rest of Owl, since the rendering pipeline is independant of the observer.

But, to come back to your post, without more details, I do not see how this would enable important features like observable primitive types and smarter memoization of store getters (what do you mean by 'smarter'? We can memoize store getters right now).

**Comment by alexkuhn** [comment]
Hello @ged-odoo


My previous post was not a troll, and I am sorry if it looks like one. I should have chosen better words to express my concerns over the concurrent rendering issues with the framework.

"All concurrent rendering issues" was indeed too broad and my mistake for calling it that way, as it omits a big amount of semantics stuffs. To clarify this post, consider all concurrent rendering issues caused by running some rendering code with business logic (e.g. template evaluation) with a mix of outdated and new business data.

To clarify which points are replies of yours, I have kept the same numbered points:


1. The rendering pipeline does not care of observable mutations, which is why immutability is very important. If it keeps mutations, then all faulty renderings should be aborted immediately. Maybe this could be implemented with custom "abortable" promises from the framework internals, but I think this is a worse solution.

2. Such concurrent rendering crashes like in the 2nd post of this issue come just from using OWL store and components the intended way, so the framework should ideally not have these issues. These bugs are hard to reproduce, but there are much more than what this issue suggests. I just wanted to tell you that there is a design issue with using mutable observables and asynchronous rendering. This is not just a missing guards here-and-there kind of bug.

3. `useStore()` uses a mix of rev numbers comparison, shallow equality, and (strict) equality. The former has highest precedence, and this can have weird behaviour like not triggering a re-render when replacing an observable object `A` by another object `B` (note: `newRevNumber` is non-zero and has same value as before, https://github.com/odoo/owl/blob/master/src/store.ts#L101 ) 

4. I think observable primitive type is important because a lot of observable stuffs are primitive types. Without this, we are encouraged to either define one single `useStore()`/`useState()` per component or wrap primitive values in an object with `{ value: value }`. In our code, we have made a custom hook that observes a boolean (basically determines the visibility of a component based on user actions). Wrapping the boolean in an observable object with `useState` makes our templates and business code hard to read in a quite significant way.

By "smarter" memoization of store getters, I meant that re-computation of store getters only happens when related store state have changed, instead of on all store changes at present: at the moment, store getters are memoized for a very short time, and they are computed only once per successive rendering from a single synchronous batch of store changes. They are still re-computed even when store changes are unrelated. Vuex does something like this, but their feature is quite weak and limited to just a kind of getters. OWL could have a much better implementation with store getters having dependencies over a slice of store state. This feature would take advantage of immutability to detect relevant changes, including primitive types.

Also, another problem with store getters I forgot in my previous post but quite relevant, is that usually using them in `useStore` triggers re-render whenever store has changes. This happens when getters create a new object or array internally. This is a big issue in my opinion, and the idea behind "smarter" memoization of store getters could help solve this issue.



Happy to see that you are looking at this with #294, but at the same time a bit sad this is not considered for v1. To be honest, I'm afraid more such rendering issues will happen in the future, and some of them will be so hard to solve with the approach with scenarios that having the right design could avoid all of this.

**Comment by ged-odoo** [comment]
Well, I am not really sure what to say. You are giving a bunch of opinions, but not much example/facts, it makes it hard to discuss. Please come see me whenever you come back from vacation. It will be much better to have a discussion in person. I am sure that it will be more productive.
